### PR TITLE
Jeffders/connector

### PR DESCRIPTION
**Pull request by Jeffders**
Support to allow other channel <--> bot authentication schemes. This allows customization of how the bot validates channel tokens and also how the bot generates tokens.

### CODE REVIEW

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/ConfigurationChannelProvider.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        /// <summary>
+        /// The key for ChannelService.
+        /// </summary>
+        public const string ChannelServiceKey = "ChannelService";
+        
+        public ConfigurationChannelProvider(IConfiguration configuration)
+        {
+            this.ChannelService = configuration.GetSection(ChannelServiceKey)?.Value;
+        }
```
Comment (cleemullins) [inline libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/ConfigurationChannelProvider.cs]: Should this throw if the value is not present?

File: libraries/Microsoft.Bot.Connector/Authentication/AuthenticationConstants.cs [hunk 2]
```diff
@@ -10,6 +10,11 @@ namespace Microsoft.Bot.Connector.Authentication
     /// </summary>
     public static class AuthenticationConstants
     {
+        /// <summary>
+        /// Government Channel Service property value
+        /// </summary>
+        public const string GovernmentChannelService = "https://botframework.gov";
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/AuthenticationConstants.cs]: The name here is unexpected. By having a "GovernmentChannelService" I expect to also see a "RegularChannelService" on [https://botframework.com](https://botframework.com). Likewise the ".gov" vs ".us" URIs, all with with "Government" in the name is additionally confusing. 

Is there a way to simplify this?

File: libraries/Microsoft.Bot.Connector/Authentication/GovernmentChannelValidation.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+using System.Linq;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Microsoft.Bot.Connector.Authentication
+{
+    public class GovernmentChannelValidation
+    {
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/GovernmentChannelValidation.cs]: sealed

File: libraries/Microsoft.Bot.Connector/Authentication/EnterpriseChannelValidation.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+using System.Linq;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Microsoft.Bot.Connector.Authentication
+{
+    public class EnterpriseChannelValidation
+    {
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/EnterpriseChannelValidation.cs]: sealed

File: libraries/Microsoft.Bot.Connector/Authentication/AuthenticationConstants.cs [hunk 5]
```diff
@@ -10,6 +10,11 @@ namespace Microsoft.Bot.Connector.Authentication
     /// </summary>
     public static class AuthenticationConstants
     {
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/AuthenticationConstants.cs]: I would probably make this a sealed class, so nobody can derive from it and change constants

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs [hunk 6]
```diff
@@ -29,7 +29,7 @@ public static class JwtTokenValidation
         /// <returns>A task that represents the work queued to execute.</returns>
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs]: Add the provider parameter

**Comment by cleemullins** [comment]
>                 if (isAuthDisabled)

This should - I think - have a .ConfigureAwait() at the end of it. 

---
Refers to: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:37 in 3db1548. [](commit_id = 3db15488a85f29c2fd2c7754e567af807f72d51f, deletion_comment = False)

**Comment by cleemullins** [comment]
>         /// <param name="serviceUrl">The service URL for the activity.</param>

Add channelProvider as a parameter 

---
Refers to: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs:62 in 3db1548. [](commit_id = 3db15488a85f29c2fd2c7754e567af807f72d51f, deletion_comment = False)

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs [hunk 7]
```diff
@@ -75,6 +75,17 @@ public static async Task<ClaimsIdentity> ValidateAuthHeader(string authHeader, I
             {
                 return await EmulatorValidation.AuthenticateEmulatorToken(authHeader, credentials, httpClient ?? _httpClient, channelId);
             }
+            else if(channelProvider != null && !string.IsNullOrEmpty(await channelProvider.GetChannelServiceAsync().ConfigureAwait(false)))
+            {
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs]: This line is too complex. Please simplify. Strongly consider puting the async call on it's own line, not direclty in the if statement.

File: libraries/Microsoft.Bot.Connector/Authentication/MicrosoftGovernmentAppCredentials.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    /// </summary>
+    public class MicrosoftGovernmentAppCredentials : MicrosoftAppCredentials
+    {
+        /// <summary>
+        /// An empty set of credentials.
+        /// </summary>
+        public static new readonly MicrosoftGovernmentAppCredentials Empty = new MicrosoftGovernmentAppCredentials(null, null);
+        
+        static MicrosoftGovernmentAppCredentials()
+        {
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/MicrosoftGovernmentAppCredentials.cs]: As a general rule type initializers are best avoided, as it's extremely non-deterministic when they run. 

Please strongly consider removing this type initializer and doing this a different way. 

Jon Skeet uses these as brain-teasers as the behavior is so wildly unpredictable:
[https://www.google.com/search?q=jon+skeet+type+initializer](https://www.google.com/search?q=jon+skeet+type+initializer)

File: libraries/Microsoft.Bot.Connector/Authentication/SimpleChannelProvider.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        public string ChannelService { get; set; }
+
+        public SimpleChannelProvider()
+        {
+        }
+
+        public SimpleChannelProvider(string channelService)
+        {
+            this.ChannelService = channelService;
+        }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/SimpleChannelProvider.cs]: null parameter check / throw on null

File: libraries/Microsoft.Bot.Connector/Authentication/SimpleChannelProvider.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Threading.Tasks;
+
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Connector/Authentication/SimpleChannelProvider.cs]: REmove & Sort usings (I don't think you're using LINQ anywhere)

File: tests/Microsoft.Bot.Connector.Tests/MicrosoftGovernmentAppCredentialsTests.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        [Fact]
+        public void MicrosoftGovernmentAppCredentials_Has_Gov_Endpoint()
+        {
+            var cred = new MicrosoftGovernmentAppCredentials(string.Empty, string.Empty);
+
+            Assert.Contains("login.microsoftonline.us", cred.OAuthEndpoint);
+        }
+        
+        [Fact]
+        public void MicrosoftGovernmentAppCredentials_Uses_Gov_Scope()
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Connector.Tests/MicrosoftGovernmentAppCredentialsTests.cs]: The mostly likely "silly" (or malicious) mistake is that someone is going to change one of the URI's used (framework.gov, etc). 

Consider adding a simple check here that tests each of the key constants. If these constants are changed and we don't notice (aka: no tests fail) then we'll have a problem.

File: tests/Microsoft.Bot.Connector.Tests/MicrosoftGovernmentAppCredentialsTests.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+using Xunit;
+
+namespace Microsoft.Bot.Connector.Tests
+{
+    public class MicrosoftGovernmentAppCredentialsTests
+    {
+        [Fact]
+        public void MicrosoftGovernmentAppCredentials_Has_Gov_Endpoint()
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Connector.Tests/MicrosoftGovernmentAppCredentialsTests.cs]: Given the amount of new code added, this doesn't seem like enough tests. 

Consider adding tests to cover a broader range of new behavior.

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/ConfigurationChannelProvider.cs (hunk 1 above)
Comment (Jeffders) [reply libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/ConfigurationChannelProvider.cs]: Well, blank or null means we use "regular" channel service values. There isn't an exception when there isn't an AppId or Password, right? I could throw and say "only use this if you are connecting to gov or a private DirectLine and here is the constant to use"...what do you think?

File: libraries/Microsoft.Bot.Connector/Authentication/AuthenticationConstants.cs (hunk 2 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Connector/Authentication/AuthenticationConstants.cs]: I will change the constant to have a .us ending.

I think I'll move all the government constants to its own class and same with the enterprise one.

File: libraries/Microsoft.Bot.Connector/Authentication/GovernmentChannelValidation.cs (hunk 3 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Connector/Authentication/GovernmentChannelValidation.cs]: OK.

File: libraries/Microsoft.Bot.Connector/Authentication/EnterpriseChannelValidation.cs (hunk 4 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Connector/Authentication/EnterpriseChannelValidation.cs]: OK.

File: libraries/Microsoft.Bot.Connector/Authentication/AuthenticationConstants.cs (hunk 5 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Connector/Authentication/AuthenticationConstants.cs]: Static classes cannot be sealed...you get one or the other :/

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs (hunk 6 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs]: OK.

File: libraries/Microsoft.Bot.Connector/Authentication/MicrosoftGovernmentAppCredentials.cs (hunk 8 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Connector/Authentication/MicrosoftGovernmentAppCredentials.cs]: I got rid of these and just added them to the base class. There is no harm in adding to our "known good" list.

File: libraries/Microsoft.Bot.Connector/Authentication/SimpleChannelProvider.cs (hunk 10 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Connector/Authentication/SimpleChannelProvider.cs]: Eyes like an eagle, he has!

File: tests/Microsoft.Bot.Connector.Tests/MicrosoftGovernmentAppCredentialsTests.cs (hunk 11 above)
Comment (Jeffders) [reply tests/Microsoft.Bot.Connector.Tests/MicrosoftGovernmentAppCredentialsTests.cs]: Ok, sounds good :)

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/ConfigurationChannelProvider.cs (hunk 1 above)
Comment (Jeffders) [reply libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/ConfigurationChannelProvider.cs]: No, the value can be null/empty string and that is 'Public Azure' as this is what we have now. The reason I prefer it this way is that I want people to use the same bot code for a bot deployment in Gov and non-Gov where the only change is to the app.settings. This means we have to be ok with putting nothing in there but still trying to read the property.

File: libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs (hunk 7 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Connector/Authentication/JwtTokenValidation.cs]: Clean as a whistle now.

File: libraries/Microsoft.Bot.Connector/Authentication/SimpleChannelProvider.cs (hunk 9 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Connector/Authentication/SimpleChannelProvider.cs]: No, null is an ok value. I added this in comments (and see above comment for rational).

File: tests/Microsoft.Bot.Connector.Tests/MicrosoftGovernmentAppCredentialsTests.cs (hunk 12 above)
Comment (Jeffders) [reply tests/Microsoft.Bot.Connector.Tests/MicrosoftGovernmentAppCredentialsTests.cs]: done

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/ConfigurationChannelProvider.cs (hunk 1 above)
Comment (cleemullins) [reply libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/ConfigurationChannelProvider.cs]: Can you put this into a code comment here? That way people can see it. 

---
In reply to: [214128651](https://github.com/Microsoft/botbuilder-dotnet/pull/890#discussion_r214128651) [](ancestors = 214128651)

**Pull request merged**